pkg/controller/master/node/node_down_controller.go

```diff
@@ -156,3 +174,89 @@ func getNodeCondition(conditions []corev1.NodeCondition, conditionType corev1.No
 	}
 	return cond
 }
+
+// deleteVolumeAttachmentsOfPod deletes all volume attachments of a pod,
+// when the pod is on a down node.
+// due to controller-manager handle too long time, cause VM restart wait a long time.
+// then we delete VolumeAttachments directly.
+// ref: https://github.com/harvester/harvester/issues/4049
+func (h *nodeDownHandler) deleteVolumeAttachmentsOfPod(pod *corev1.Pod) error {
+	volumeAttachments, err := h.getVolumeAttachmentsOfPod(pod)
+	if err != nil {
+		return err
+	}
+	for _, va := range volumeAttachments {
+		if va.DeletionTimestamp == nil {
+			err := h.vas.Delete(va.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				if errors.IsNotFound(err) {
+					continue
+				}
+				return err
+			}
+			logrus.Infof("deleted volume attachment %v for pod %v on downed node %v",
+				va.Name,
+				pod.Name,
+				pod.Spec.NodeName)
+		}
+	}
+	return nil
+}
+
+func (h *nodeDownHandler) getVolumeAttachmentsOfPod(pod *corev1.Pod) ([]*storagev1.VolumeAttachment, error) {
+	res := make([]*storagev1.VolumeAttachment, 0)
+	volumeAttachments, err := h.vaCache.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	pvs := make(map[string]bool)
+
+	for _, vol := range pod.Spec.Volumes {
+		if vol.VolumeSource.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := h.pvcCache.Get(pod.Namespace, vol.VolumeSource.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				continue
+			}
+			return nil, err
+		}
+		pvs[pvc.Spec.VolumeName] = true
+	}
+
+	for _, va := range volumeAttachments {
+		if va.Spec.NodeName != pod.Spec.NodeName {
+			continue
+		}
+		if va.Spec.Attacher != longhorntypes.LonghornDriverName {
+			continue
+		}
+		if va.Spec.Source.PersistentVolumeName == nil {
+			continue
+		}
+		if _, ok := pvs[*va.Spec.Source.PersistentVolumeName]; !ok {
+			continue
+		}
+		res = append(res, va)
+	}
+
+	return res, nil
+}
+
+// resetHeartbeat force reduce node's heartbeat to immediately rebuild VMI,
+// Ensure that can take effect here by reducing 6 minutes
+// ref: https://kubevirt.io/user-guide/operations/unresponsive_nodes/#virt-handler-heartbeat
+func (h *nodeDownHandler) resetHeartbeat(node *corev1.Node) (*corev1.Node, error) {
+	nodeCpy := node.DeepCopy()
+	if nodeCpy.Annotations == nil {
+		nodeCpy.Annotations = map[string]string{}
+	}
+
+	t := metav1.Now().Add(-6 * time.Minute).Format(time.RFC3339)
+	nodeCpy.Annotations[kubevirtv1.VirtHandlerHeartbeat] = t
+
+	return h.nodes.Update(nodeCpy)
```

Comment:
Should the update be checked with DeepEqual? In my environment, the heartbeat kept renewing.